moveit_ros/moveit_servo/src/servo.cpp

```diff
@@ -513,9 +538,75 @@ KinematicState Servo::getNextJointState(const ServoInput& command)
     }
   }
 
+  // remove old commands
+  cur_time = node_->now();
+  while (!committed_commands_.empty() &&
+         committed_commands_.front().time < (cur_time - rclcpp::Duration::from_seconds(2 * servo_params_.latency)))
+  {
+    committed_commands_.pop_front();
+  }
+
+  // add next committed command
+  target_state.time = node_->now() + rclcpp::Duration::from_seconds(servo_params_.latency);
+  committed_commands_.push_back(target_state);
+
+  // add end command stop point in case of large delay
+  auto dt = 2 * (servo_params_.latency);
+  auto end_state = target_state;
+  for (int i = 0; i < num_joints; i++)
+  {
+    end_state.positions[i] = target_state.positions[i] + target_state.velocities[i] * dt;
+    end_state.velocities[i] = 0;
+    end_state.accelerations[i] = 0;
+    end_state.time = target_state.time + rclcpp::Duration::from_seconds(dt);
+  }
+  committed_commands_.push_back(end_state);
+
   return target_state;
 }
 
+trajectory_msgs::msg::JointTrajectory Servo::createTrajectoryMessage()
+{
+  // TODO should have realtime capabilities
+  trajectory_msgs::msg::JointTrajectory joint_trajectory;
+
+  // remove old commands
+  auto cur_time = node_->now();
+  while (!committed_commands_.empty() &&
+         committed_commands_.front().time < (cur_time - rclcpp::Duration::from_seconds(2 * servo_params_.latency)))
+  {
+    committed_commands_.pop_front();
+  }
+  if (committed_commands_.empty())
+  {
+    RCLCPP_ERROR(node_->get_logger(),
+                 "Command queue is empty when creating trajectory, the newest command is already in the past!");
+    return joint_trajectory;
+  }
+  joint_trajectory.joint_names = committed_commands_.front().joint_names;
+  joint_trajectory.points.reserve(committed_commands_.size());
+  joint_trajectory.header.stamp = committed_commands_.front().time;
+  for (const auto& state : committed_commands_)
+  {
+    trajectory_msgs::msg::JointTrajectoryPoint point;
+    size_t num_joints = state.positions.size();
+    point.positions.reserve(num_joints);
+    point.velocities.reserve(num_joints);
+    for (const auto& pos : state.positions)
+    {
+      point.positions.emplace_back(pos);
+    }
+    for (const auto& vel : state.velocities)
+    {
+      point.velocities.emplace_back(vel);
+    }
+    point.time_from_start = state.time - joint_trajectory.header.stamp;
+    joint_trajectory.points.emplace_back(point);
```

Comment:
Refer to the `composeTrajectoryMessage()` function in `common.cpp`, but I believe there are parameters such as `servo_params.publish_joint_positions/velocities/accelerations` that control what data to actually pack into the `JointTrajectory` message. This should be consistent with that.